src/backend/distributed/commands/extension.c

```diff
@@ -115,3 +212,303 @@ ExtractNewExtensionVersion(Node *parsetree)
 
 	return newVersion;
 }
+
+
+#if PG_VERSION_NUM >= 100000
+static Datum
+DirectFunctionCall0Coll(PGFunction func, Oid collation)
+{
+	FunctionCallInfoData fcinfo;
+	Datum result;
+
+	InitFunctionCallInfoData(fcinfo, NULL, 0, collation, NULL, NULL);
+
+	result = (*func)(&fcinfo);
+
+	/* Check for null result, since caller is clearly not expecting one */
+	if (fcinfo.isnull)
+	{
+		elog(ERROR, "function %p returned NULL", (void *) func);
+	}
+
+	return result;
+}
+
+
+#endif
+
+
+static bool
+ShouldUseAutoSSL(void)
+{
+	const char *sslmode = NULL;
+	sslmode = GetConnParam("sslmode");
+
+	if (strcmp(sslmode, "require") == 0)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+#ifdef USE_SSL
+
+/*
+ * CreateCertificatesWhenNeeded checks if the certificates exists. When they don't exist
+ * they will be created. The return value tells whether or not new certificates have been
+ * created. After this function it is guaranteed that certificates are in place. It is not
+ * guaranteed they have the right permissions as we will not touch the keys if they exist.
+ */
+static bool
+CreateCertificatesWhenNeeded()
+{
+	EVP_PKEY *pkey = NULL;
+	X509 *x509 = NULL;
+	bool filesWritten = false;
+	bool useDSA = false;
+
+	/*
+	 * check if we can load the certificate, when we can we assume the certificates are im
+	 * place
+	 */
+	if (SSL_CTX_use_certificate_chain_file(NULL, ssl_cert_file) == 1)
+	{
+		/* we didn't create any so return false */
+		return false;
+	}
+
+	elog(LOG, "no certificate present, generating self signed certificate");
+
+	/* Generate the key. */
+	if (useDSA)
+	{
+		pkey = generate_key_dsa();
+	}
+	else
+	{
+		pkey = generate_key_rsa();
+	}
+
+	if (!pkey)
+	{
+		ereport(ERROR, (errmsg("error while generating private key")));
+		return false;
+	}
+
+	/* Generate the certificate. */
+	x509 = generate_x509(pkey);
+	if (!x509)
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("error while generating x509 certificate")));
+		return false;
+	}
+
+	/* Write the private key and certificate out to disk. */
+	filesWritten = write_to_disk(pkey, x509);
+	EVP_PKEY_free(pkey);
+	X509_free(x509);
+
+	if (!filesWritten)
+	{
+		ereport(ERROR, (errmsg("error while writing key and certificate to file")));
+		return false;
+	}
+
+	return true;
+}
+
+
+/* Generates a 2048-bit RSA key. */
+static EVP_PKEY *
+generate_key_rsa()
+{
+	int ret = 0;
+	EVP_PKEY *pkey = NULL;
+	BIGNUM *bne = NULL;
+	RSA *rsa = NULL;
+
+	/* Allocate memory for the EVP_PKEY structure. */
+	pkey = EVP_PKEY_new();
+	if (!pkey)
+	{
+		ereport(ERROR, (errmsg("unable to create EVP_PLEY")));
+		return NULL;
+	}
+
+	/* 1. generate rsa key */
+	bne = BN_new();
+	ret = BN_set_word(bne, RSA_F4);
+	if (ret != 1)
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("unable to prepare bignum")));
+		return NULL;
+	}
+
+	/* Generate the RSA key and assign it to pkey. */
+	rsa = RSA_new();
+	ret = RSA_generate_key_ex(rsa, 2048, bne, NULL);
```

Comment:
do we need to seed the PRNG using `RAND_poll` prior to this call? (e.g. like in `pg_strong_random`)

Reply:
seems this is done automatically, openssl docs are a bit confusing

Reply:
These are the exact things we need to have more eyes on. From what I could find is that this is taken care of, however as you found the docs are not clear around these important topics.